In [1]:
import pandas as pd
import numpy as np

In [7]:
raw_data = pd.read_csv('Q4_ADX_Sample_data.csv')
raw_data.head(10)

date  pub_id    app_id  ad_unit_code         country  requests  \
0  2022-10-20    1987  16243270   22832846443        Tanzania        36   
1  2022-10-20    1987  16243270   22832846593        Tanzania        24   
2  2022-10-20    1987  16243270   22832846443        Thailand       949   
3  2022-10-20    1987  16243270   22832846593        Thailand      1006   
4  2022-10-20    1987  16243270   22832827729        Thailand       373   
5  2022-10-20    1987  16243270   22832846443   United States       225   
6  2022-10-20    1987  16243270   22832827729   United States        39   
7  2022-10-20    1987  16243270   22832846593   United States       189   
8  2022-10-20    1987  16243270   22832846593  United Kingdom        33   
9  2022-10-20    1987  16243270   22832846443  United Kingdom        37   

   ae_responses  ae_impressions  ae_clicks  ae_revenue  
0            36              34          0    0.035738  
1            24              23          0    0.064384  
2           949             876         20    4.943878  
3           907             688        134   24.226836  
4           287             156         20    2.759392  
5           225             212          2    4.316628  
6            39              30          2    1.450446  
7           189             165          4   20.327859  
8            33              25          1    1.874777  
9            37              36          1    0.269845

In [13]:
print(raw_data.shape)

(47935, 10)


In [8]:
for col in raw_data.columns:
    print(col)

date
pub_id
app_id
ad_unit_code
country
requests
ae_responses
ae_impressions
ae_clicks
ae_revenue


In [20]:
requests_check = raw_data[['pub_id','app_id','requests','ae_responses','ae_impressions','ae_clicks','ae_revenue']].groupby(['pub_id','app_id']).sum()
requests_check['show_rate'] = requests_check['ae_impressions']/requests_check['ae_responses']
requests_check['click_rate'] = requests_check['ae_clicks']/requests_check['ae_impressions']
requests_check['engangement_revenue'] = requests_check['ae_revenue']/requests_check['ae_clicks']
requests_check

requests  ae_responses  ae_impressions  ae_clicks  \
pub_id app_id                                                        
14     22693095  10349859       9459787         3524462     359318   
       95674771   1211485       1182134          754469      35976   
34     66981243    998870        964582          676995      41647   
51     47862278   5965420       4674814         3150826      27628   
72     84650721    853355        824285          650202      80681   
107    63338696   2845293       1975614         1124600      40869   
281    40931528   2612285       2533914         1096917      15626   
879    25596802     86944         85854           71831       7901   
1987   16243270    144079        140361          113944       2703   
2808   43840611   3014066       2937129         1157692      34894   

                   ae_revenue  show_rate  click_rate  engangement_revenue  
pub_id app_id                                                              
14     22693095  51191.627809   0.372573    0.101950             0.142469  
       95674771   2868.401602   0.638226    0.047684             0.079731  
34     66981243   2077.719197   0.701853    0.061517             0.049889  
51     47862278   3191.668051   0.674000    0.008768             0.115523  
72     84650721   2387.244188   0.788807    0.124086             0.029589  
107    63338696   3269.125146   0.569241    0.036341             0.079990  
281    40931528  21900.482256   0.432894    0.014245             1.401541  
879    25596802   4077.663850   0.836665    0.109994             0.516095  
1987   16243270   4245.317619   0.811792    0.023722             1.570595  
2808   43840611   5373.715967   0.394158    0.030141             0.154001

## Anomalies Observed

### 1.	25596802 served the least amount of impressions (71831) with a successful response rate of 98.7% for a response for the ads requested. This app is the most efficient in generating revenues for ADX as their ad serves 0.11% CTR (the average for all the apps would be 0.6%). Although it generated a revenue of USD4077.6, it earned an average of USD57 in revenue for each 1000 impressions served, the highest from all the apps. Great potential lies in this app to generate high volumes of revenue for ADX.
### 2.	47862278 has the least CTR (0.01%) which would not , generating only USD1.0 for each 1000 impressions served. Even though it’s the second highest in terms of the quantity of impressions, 3150826, it is one of the least revenue-generating-potential apps from the sample set and needs a new strategy.